### DISCUSSION TITLE
pre-commit manager skips ssh repo with invalid-url

### DISCUSSION THREAD

**Discussion by wwuck** [renovatebot/renovate#24798] [Suggest an Idea]
### How are you running Renovate?

Self-hosted

### Please select which platform you are using if self-hosting.

Bitbucket Server

### If you're self-hosting Renovate, tell us what version of Renovate you run.

Docker renovate/renovate:26.13.8

### Describe the bug

I have a repository (`RT/repo2`) with a `.pre-commit-config.yaml` file that contains a pre-commit hook repo (`PREC/pre-commit`) hosted on my Bitbucket Server instance, the same server I run renovate-bot on.  I have Bitbucket Server set to only use ssh for repository cloning.  I mount the ssh keys inside the Docker container and there are no problems with renovate cloning the repository `RT/repo2`.

```yaml
# See https://pre-commit.com for more information
# See https://pre-commit.com/hooks.html for more hooks
---
default_language_version:
  python: python3.9
repos:
  - repo: ssh://git@bitbucketserver.example.com:7999/prec/pre-commit.git
    rev: v1.2.0   # v1.3.0 is the latest tag in the remote repo
    hooks:
      - id: shellcheck-docker
      - id: shfmt-docker

```

When I run renovate, it fails to detect the dependency update and gives the error `INFO: Could not separate hostname from full dependency url. (repository=ssh://git@bitbucketserver.example.com:7999/prec/pre-commit.git)`

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Matched 1 file(s) for manager pre-commit: .pre-commit-config.yaml (repository=RT/repo2)
DEBUG: Found version (repository=RT/repo2)
       "tag": "v1.2.0"
 INFO: Could not separate hostname from full dependency url. (repository=ssh://git@bitbucketserver.example.com:7999/prec/pre-commit.git)
DEBUG: Found pre-commit package files (repository=RT/repo2)
DEBUG: Found 1 package file(s) (repository=RT/repo2)
 INFO: Dependency extraction complete (repository=RT/repo2)
       "baseBranch": "main",
       "stats": {
         "managers": {"pre-commit": {"fileCount": 1, "depCount": 1}},
         "total": {"fileCount": 1, "depCount": 1}
       }
DEBUG: Package releases lookups complete (repository=RT/repo2)
       "baseBranch": "main"
DEBUG: branchifyUpgrades (repository=RT/repo2)
DEBUG: 0 flattened updates found:  (repository=RT/repo2)
DEBUG: Returning 0 branch(es) (repository=RT/repo2)
DEBUG: config.repoIsOnboarded=false (repository=RT/repo2)
DEBUG: packageFiles with updates (repository=RT/repo2)
       "config": {
         "pre-commit": [
           {
             "packageFile": ".pre-commit-config.yaml",
             "deps": [
               {
                 "depType": "repository",
                 "skipReason": "invalid-url",
                 "currentValue": "v1.2.0",
                 "depIndex": 0,
                 "updates": []
               }
             ]
           }
         ]
       }
DEBUG: ensureOnboardingPr() (repository=RT/repo2)
DEBUG: getBranchPr(renovate/configure) (repository=RT/repo2)
DEBUG: findPr(renovate/configure, "undefined", "open") (repository=RT/repo2)
DEBUG: getPrList() (repository=RT/repo2)
DEBUG: returning cached PR list (repository=RT/repo2)
DEBUG: Found PR #1 (repository=RT/repo2)
DEBUG: getPr(1) (repository=RT/repo2)
DEBUG: Filling in onboarding PR template (repository=RT/repo2)
DEBUG: getConfigDesc() (repository=RT/repo2)
DEBUG: getDescriptionArray() (repository=RT/repo2)
DEBUG: getScheduleDesc() (repository=RT/repo2)
DEBUG: No schedule (repository=RT/repo2)
DEBUG: Found description array (repository=RT/repo2)
       "length": 20
DEBUG: getPrList() (repository=RT/repo2)
DEBUG: massageMarkdown(Welcome to [Renovate](https://github.com/renovatebot/renovate)! This is an onboarding PR to help you understand and configure settings before regular Pull Requests begin.) (repository=RT/repo2)
DEBUG: Found open onboarding PR (repository=RT/repo2)
DEBUG: Pull Request #1 does not need updating (repository=RT/repo2)
DEBUG: processRepo() (repository=RT/repo2)
DEBUG: Repo is onboarding - skipping dependency dashboard (repository=RT/repo2)
DEBUG: Removing any stale branches (repository=RT/repo2)
DEBUG: config.repoIsOnboarded=false (repository=RT/repo2)
DEBUG: Branch lists (repository=RT/repo2)
       "branchList": ["renovate/configure"],
       "renovateBranches": ["renovate/configure"]
DEBUG: remainingBranches= (repository=RT/repo2)
DEBUG: No branches to clean up (repository=RT/repo2)
DEBUG: ensureIssueClosing(Action Required: Fix Renovate Configuration) (repository=RT/repo2)
DEBUG: Repository timing splits (milliseconds) (repository=RT/repo2)
       "splits": {"init": 3060, "extract": 363, "lookup": 5, "update": 15},
       "total": 3448
DEBUG: http statistics (repository=RT/repo2)
       "hostStats": {"bitbucketserver.example.com": {"requestCount": 7, "requestAvgMs": 64, "queueAvgMs": 0}},
       "totalRequests": 7
 INFO: Repository finished (repository=RT/repo2)
       "durationMs": 3448
DEBUG: Renovate exiting
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by rarkins** [comment]
Renovate currently supports lookups over https only

**Comment by viceice** [comment]
BTW: this looks ugly too:
https://github.com/renovatebot/renovate/blob/f0297209ec3e49921643040e399351e3ea41cbe3/lib/manager/pre-commit/extract.ts#L58

gitea -> gitlab-tags ?

**Comment by viceice** [comment]
On small bug is there, we don't support ports in hostname, eg `https://gitlab.domain.com:7777/user/repo.git` won't work too.

https://github.com/renovatebot/renovate/blob/f0297209ec3e49921643040e399351e3ea41cbe3/lib/manager/pre-commit/extract.ts#L90-L97

So we should at least allow the port in hostname

**Comment by wwuck** [comment]
Thanks.  I guess this issue is now a feature request for lookups over ssh.

**Comment by rarkins** [comment]
@wwuck questions about your use case:
- Is this the same server/host as your git is hosted on? (e.g. corresponds with `RENOVATE_ENDPOINT` value)
- Is Renovate cloning over git for repos too, or able to use https?

**Comment by wwuck** [comment]
The pre-commit hooks repository is hosted on the same bitbucket server as my renovate endpoint.

I have the bitbucket server SCM connection settings set to `HTTP(S) disabled` and `SSH enabled` with the default ssh port of `7999` and also `SSH access keys enabled`.  All ssh access is done via ssh access keys so there are no usernames/passwords needed for the clone.  This results in all cloning being done over SSH.

The clone url for cloning over SSH when I click on the `clone` button in a bitbucket server repository is in the format of `ssh://git@bitbucketserver.example.com:7999/prec/pre-commit.git`.  I can see this is different to the github.com SSH clone url which doesn't include the `ssh://` prefix.

The bitbucket documentation page for SSH access is https://confluence.atlassian.com/bitbucketserver/enable-ssh-access-to-git-repositories-776640358.html.  Also relevant is https://confluence.atlassian.com/bitbucketserver/disable-http-s-access-to-git-repositories-776640365.html.
